--- a/networking/udhcp/common.c	2022-06-28 09:32:48.853072914 +0800
+++ b/networking/udhcp/common.c	2022-06-28 09:39:28.000000000 +0800
@@ -54,6 +54,7 @@
 	{ OPTION_STRING                           , 0x43 }, /* DHCP_BOOT_FILE     */
 //TODO: not a string, but a set of LASCII strings:
 //	{ OPTION_STRING                           , 0x4D }, /* DHCP_USER_CLASS    */
+	{ OPTION_IP	| OPTION_LIST				  , 0x8A }, /* DHCP_CAPWAP       */
 	{ OPTION_STRING                           , 0x64 }, /* DHCP_PCODE         */
 	{ OPTION_STRING                           , 0x65 }, /* DHCP_TCODE         */
 #if ENABLE_FEATURE_UDHCP_RFC3397
@@ -123,6 +124,7 @@
 	"tftp" "\0"             /* DHCP_TFTP_SERVER_NAME*/
 	"bootfile" "\0"         /* DHCP_BOOT_FILE       */
 //	"userclass" "\0"        /* DHCP_USER_CLASS      */
+	"capwap" "\0"      		/* DHCP_CAPWAP		*/
 	"tzstr" "\0"            /* DHCP_PCODE           */
 	"tzdbstr" "\0"          /* DHCP_TCODE           */
 #if ENABLE_FEATURE_UDHCP_RFC3397
