Index: netifd-2021-05-26-1eb0fafa/system-linux.c
===================================================================
--- netifd-2021-05-26-1eb0fafa.orig/system-linux.c
+++ netifd-2021-05-26-1eb0fafa/system-linux.c
@@ -1578,6 +1578,8 @@ int system_vlandev_del(struct device *vl
 	return system_link_del(vlandev->ifname);
 }
 
+#ifdef SPEED_400000
+// ugly hack to detect if we are running on an ancient kernel
 static void
 system_set_ethtool_settings(struct device *dev, struct device_settings *s)
 {
@@ -1628,6 +1630,7 @@ system_set_ethtool_settings(struct devic
 	ecmd.cmd = ETHTOOL_SSET;
 	ioctl(sock_ioctl, SIOCETHTOOL, &ifr);
 }
+#endif
 
 void
 system_if_get_settings(struct device *dev, struct device_settings *s)
@@ -1852,7 +1855,9 @@ system_if_apply_settings(struct device *
 		system_set_drop_unsolicited_na(dev, s->drop_unsolicited_na ? "1" : "0");
 	if (apply_mask & DEV_OPT_ARP_ACCEPT)
 		system_set_arp_accept(dev, s->arp_accept ? "1" : "0");
+#ifdef SPEED_400000
 	system_set_ethtool_settings(dev, s);
+#endif
 }
 
 int system_if_up(struct device *dev)
