--- a/drivers/hwmon/adt7470.c
+++ b/drivers/hwmon/adt7470.c
@@ -1271,10 +1271,17 @@ static const struct i2c_device_id adt747
 };
 MODULE_DEVICE_TABLE(i2c, adt7470_id);
 
+static const struct of_device_id __maybe_unused adt7470_of_match =
+{
+	.compatible = "adi,adt7470",
+};
+MODULE_DEVICE_TABLE(of, adt7470_of_match);
+
 static struct i2c_driver adt7470_driver = {
 	.class		= I2C_CLASS_HWMON,
 	.driver = {
 		.name	= "adt7470",
+		.of_match_table = of_match_ptr(&adt7470_of_match),
 	},
 	.probe		= adt7470_probe,
 	.remove		= adt7470_remove,
